File: src/Server/Position.cs
namespace Belin.UI;

using System.Text.Json.Serialization;

/// <summary>
/// Defines the position of an element.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum Position {

	/// <summary>
	/// Top left.
	/// </summary>
	TopStart,

	/// <summary>
	/// Top center.
	/// </summary>
	TopCenter,

	/// <summary>
	/// Top right.
	/// </summary>
	TopEnd,

	/// <summary>
	/// Middle left.
	/// </summary>
	MiddleStart,

	/// <summary>
	/// Middle center.
	/// </summary>
	MiddleCenter,

	/// <summary>
	/// Middle right.
	/// </summary>
	MiddleEnd,

	/// <summary>
	/// Bottom left.
	/// </summary>
	BottomStart,

	/// <summary>
	/// Bottom center.
	/// </summary>
	BottomCenter,

	/// <summary>
	/// Bottom right.
	/// </summary>
	BottomEnd
}

/// <summary>
/// Provides extension members for element positions.
/// </summary>
public static class PositionExtensions {
	extension(Position position) {

		/// <summary>
		/// The CSS class of this position.
		/// </summary>
		public string CssClass => position switch {
			Position.TopStart => "top-0 start-0",
			Position.TopCenter => "top-0 start-50 translate-middle-x",
			Position.TopEnd => "top-0 end-0",
			Position.MiddleStart => "top-50 start-0 translate-middle-y",
			Position.MiddleCenter => "top-50 start-50 translate-middle",
			Position.MiddleEnd => "top-50 end-0 translate-middle-y",
			Position.BottomStart => "bottom-0 start-0",
			Position.BottomCenter => "bottom-0 start-50 translate-middle-x",
			Position.BottomEnd => "bottom-0 end-0"
		};
	}
}

File: Sources/Client/Position.ts
/**
 * Defines the position of an element.
 */
export const Position = Object.freeze({

	/**
	 * Top left.
	 */
	TopStart: "TopStart",

	/**
	 * Top center.
	 */
	TopCenter: "TopCenter",

	/**
	 * Top right.
	 */
	TopEnd: "TopEnd",

	/**
	 * Middle left.
	 */
	MiddleStart: "MiddleStart",

	/**
	 * Middle center.
	 */
	MiddleCenter: "MiddleCenter",

	/**
	 * Middle right.
	 */
	MiddleEnd: "MiddleEnd",

	/**
	 * Bottom left.
	 */
	BottomStart: "BottomStart",

	/**
	 * Bottom center.
	 */
	BottomCenter: "BottomCenter",

	/**
	 * Bottom right.
	 */
	BottomEnd: "BottomEnd"
});

/**
 * Defines the placement of an element.
 */
export type Position = typeof Position[keyof typeof Position];

/**
 * Returns the CSS class of the specified position.
 * @param position The position.
 * @returns The CSS class of the specified position.
 */
export function cssClass(position: Position): string {
	switch (position) {
		case Position.TopCenter: return "top-0 start-50 translate-middle-x";
		case Position.TopEnd: return "top-0 end-0";
		case Position.MiddleStart: return "top-50 start-0 translate-middle-y";
		case Position.MiddleCenter: return "top-50 start-50 translate-middle";
		case Position.MiddleEnd: return "top-50 end-0 translate-middle-y";
		case Position.BottomStart: return "bottom-0 start-0";
		case Position.BottomCenter: return "bottom-0 start-50 translate-middle-x";
		case Position.BottomEnd: return "bottom-0 end-0";
		default: return "top-0 start-0";
	}
}
